spec/justified.tla

```diff
@@ -0,0 +1,209 @@
+\* @type: ($signedVoteMessage, $checkpoint, $commonNodeState) => Bool;
+IsVoteInSupportAssumingJustifiedSource(vote, checkpoint, node_state) ==
+    \* TODO: If we end up always doing 1-state model checking, a lot of sub-checks in 
+    \* valid_FFG_vote are already a part of the state validity predicate, and couldbe omitted here 
+    \* to simplify computation
+    /\ valid_FFG_vote(vote, node_state) 
+    /\ vote.message.ffg_target.chkp_slot = checkpoint.chkp_slot
+    \* TODO: If we end up always doing 1-state model checking, we could 
+    \* precompute the relation is_ancestor_descendant_relationship(a,b), for all blocks
+    \* ahead of time as a _function_, s.t. this lookup here becomes a simple access, instead of each of them being
+    \* another pseudo-recursive computation
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        get_block_from_hash(vote.message.ffg_target.block_hash, node_state),
+        node_state
+        )
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(vote.message.ffg_source.block_hash, node_state),
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        node_state
+        )
+
+\* With the predicate, we can filter out the votes that are relevant in one step
+\* @type: ($checkpoint, $commonNodeState) => Set($signedVoteMessage);
+VotesRelevantForOneStepIteration(checkpoint, node_state) == 
+    { vote \in node_state.view_votes: IsVoteInSupportAssumingJustifiedSource(vote, checkpoint, node_state)}
+
+\* Using VotesRelevantForOneStepIteration, we can define, for every initial checkpoint C, 
+\* a sequence of maps M_i. We will refer to the domain elements of M_i as "targets" for step i.
+\* The first map M_1 for C is simply [ c \in {C} |-> VotesRelevantForOneStepIteration(C, ...) ].
+\* Each subsequent map M_{i+1} is defined in the following way:
+\* M_{i+1} ==
+\*     LET newTargets == UNION { Sources(M_i[target]): target \in DOMAIN M_i } 
+\*     IN [ target \in newTargets |-> VotesRelevantForOneStepIteration(target, ...) ]
+ 
+\* This construction is guaranteed to be finite, i.e., there exists some n, s.t. M_k is empty for all k >= n and nonempty
+\* for all 1 < i < n. The reason is as follows:
+\*   1. valid_FFG_vote(vote) requires `vote.message.ffg_source.chkp_slot < vote.message.ffg_target.chkp_slot`
+\*   2. IsVoteInSupportAssumingJustifiedSource requires `vote.message.ffg_target.chkp_slot = checkpoint.chkp_slot`
+\* Therefore, for any checkpoint C, all votes in VotesRelevantForOneStepIteration(C, ..)
+\* have sources with a strictly lower slot number. 
+\* If we take N_i to be the maximal slot number of any sorce of a vote in a set in the codomain of M_i, then
+\* we can easily see that N_i > N_{i+1}.
+\* Since the checkpoint with the lowest slot number is the genesis checkpoint C_G, and 
+\* VotesRelevantForOneStepIteration(C_G, ...) = {}, eventually S_m will be empty.
```

Comment:
Ah, this is from the old version with sets instead of maps, must have snuck through. This should read `M_k` (I also changed the index away from `m` since maps are now `M`s instead of sets `S`s)